core/src/main/java/com/linecorp/armeria/server/annotation/Decorator.java

```diff
@@ -35,11 +36,17 @@
 public @interface Decorator {
 
     /**
-     * {@link DecoratingHttpServiceFunction} implementation type. The specified class must have an accessible
-     * default constructor.
+     * {@link DecoratingHttpServiceFunction} implementation type. The specified class must either have
+     * an accessible default constructor or get injected by {@link DependencyInjector}
+     * depending on its {@link #mode()}.
      */
     Class<? extends DecoratingHttpServiceFunction> value();
 
+    /**
+     * The instance {@link CreationMode} of {@link DecoratingHttpServiceFunction} specified in {@link #value()}.
+     */
+    CreationMode mode() default CreationMode.REFLECTION;
```

Comment:
Question:

If the default mode is `REFLECTION`, this will introduce a rather big breaking change in that .

1. Users are already using dependency injection for class A
2. If we're lucky, an exception will be raised due to a missing default constructor.
3. If we're not lucky and the class has a default constructor, an object created by reflection will be used instead of the specified bean in the dependency injector.

> Because most dependency injection frameworks raise exceptions if no bean is registered. If ReflectiveDependencyInjecton created objects implicitly, it would cause another confusion. Some people would report an issue like

> NoSuchBeanDefinitionException was ignored and an unknown object was created instead.

https://github.com/line/armeria/pull/4534#discussion_r1030325103

It seems like the current change creates the above situation.

I'm wondering if we want to add a `CreationMode.DEFAULT` which checks the dependencyInjector first, and then proceeds with reflection.

Reply:
Thanks for reviewing the draft PR! 😄

Yeah, as you've pointed out, having `CreationMode.REFLECTION` as default will introduce a breaking change and may silently break an app if `DependencyInjector` is being used, but I thought there's a room for this change because `DependencyInjector` is `@UnstableApi`.


> I'm wondering if we want to add a CreationMode.DEFAULT which checks the dependencyInjector first, and then proceeds with reflection.

IIRC we wanted to avoid silent fallback to reflection when `DependencyInjector` is introduced as @ikhook stated in the same comment you referred to, it sounds like the suggested behavior of `CreationMode.DEFAULT` conflicts with this direction?

Reply:
I believe that this is an inevitable breaking change. It would be useful to fall back to `REFLECTION` if there is no injection but I'm not sure we could name it `DEFAULT`. 
If we decide to add the feature, I prefer to give it a more explicit name such as `OPTIONAL` or `REFLECTION_ON_MISSING`, and make it non-default.